File: 852. Peak Index in a Mountain Array.java
class Solution {
    public int peakIndexInMountainArray(int[] A) {
        int lo = 0, hi = A.length - 1;
        while (lo < hi) {
            int mid = (lo + hi)/2;
            if (A[mid+1] > A[mid])
                lo = mid + 1;
            else 
                hi = mid;
        }
        return lo;
        
        // for (int i = 0; i < A.length - 1; i++) {
        //     if (A[i] > A[i+1]) return i;
        // }
        // return -1;
    }
}

File: amazon/python/852. Peak Index in a Mountain Array.py
class Solution(object):
    def peakIndexInMountainArray(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        l = 0
        r = len(A) - 1
        while l <= r:
            mid = l + (r-l)//2
            if A[mid] > A[mid - 1] and A[mid] > A[mid + 1]:
                return mid
            elif A[mid] > A[mid - 1] and A[mid] < A[mid + 1]:
                # ascending
                l = mid + 1
            elif A[mid] < A[mid - 1] and A[mid] > A[mid + 1]:
                # descending 
                r = mid - 1
                # boundings 
        
        # time: O(logn)
        # space: O(1)
        # mid cannot be 0 or last because the definition 
